pkg/webhook/admission/decode.go

```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package admission
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+)
+
+// Decoder is used to decode AdmissionRequest.
+type Decoder interface {
+	// Decode decodes the raw byte object from the AdmissionRequest to the passed-in runtime.Object.
+	Decode(Request, runtime.Object) error
+}
+
+// DecodeFunc is a function that implements the Decoder interface.
+type DecodeFunc func(Request, runtime.Object) error
+
+var _ Decoder = DecodeFunc(nil)
+
+// Decode implements the Decoder interface.
+func (f DecodeFunc) Decode(req Request, obj runtime.Object) error {
+	return f(req, obj)
+}
+
+// DecodeRequest decodes the raw byte object from the admissionv1beta1.AdmissionRequest to the passed-in runtime.Object.
+func DecodeRequest(req Request, obj runtime.Object) error {
+	raw := req.AdmissionRequest.Object.Raw
+	_, _, err := codecs.UniversalDeserializer().Decode(raw, nil, obj)
+	return err
+}
+
+type decoder struct {
+	codecs serializer.CodecFactory
+}
+
+// NewDecoder creates a Decoder given the runtime.Scheme
+func NewDecoder(scheme *runtime.Scheme) (Decoder, error) {
+	return decoder{codecs: serializer.NewCodecFactory(scheme)}, nil
+}
+
+func (d decoder) Decode(req Request, obj runtime.Object) error {
+	raw := req.AdmissionRequest.Object.Raw
+	_, _, err := d.codecs.UniversalDeserializer().Decode(raw, nil, obj)
```

Comment:
generally looks good.  Might want to use the `runtime.DecodeInto(d.codecs.UniversalDeserializer(), obj)` helper, since it double-checks that the output object is the same as the input object.

Reply:
Done